pkg/dcgmexporter/system_info.go

```diff
@@ -618,87 +647,92 @@ func AddAllLinks(sysInfo SystemInfo) []MonitoringInfo {
 	return monitoring
 }
 
-func IsSwitchWatched(switchId uint, sysInfo SystemInfo) bool {
+func IsSwitchWatched(switchID uint, sysInfo SystemInfo) bool {
 	if sysInfo.sOpt.Flex {
 		return true
 	}
 
-	if len(sysInfo.sOpt.MajorRange) <= 0 {
+	// When MajorRange contains -1 value, we do monitorig of all switches
+	if len(sysInfo.sOpt.MajorRange) > 0 && sysInfo.sOpt.MajorRange[0] == -1 {
 		return true
 	}
 
-	for _, sw := range sysInfo.sOpt.MajorRange {
-		if uint(sw) == switchId {
-			return true
-		}
-
-	}
-	return false
+	return slices.ContainsFunc(sysInfo.sOpt.MajorRange, func(id int) bool {
+		return uint(id) == switchID
+	})
 }
 
-func IsLinkWatched(linkId uint, switchId uint, sysInfo SystemInfo) bool {
+func IsLinkWatched(linkIndex uint, switchID uint, sysInfo SystemInfo) bool {
 	if sysInfo.sOpt.Flex {
 		return true
 	}
 
-	for _, sw := range sysInfo.Switches {
-		if !IsSwitchWatched(sw.EntityId, sysInfo) {
-			return false
-		}
+	// Find a switch
+	switchIdx := slices.IndexFunc(sysInfo.Switches, func(si SwitchInfo) bool {
+		return IsSwitchWatched(si.EntityId, sysInfo) && si.EntityId == switchID
+	})
+
+	if switchIdx > -1 {
```

Comment:
The logic of the function is the following: 

When sysInfo.sOpt.Flex == true - we watch all links on all switches;

Otherwise: 

1. We check if the switch associated with the monitored link is watched. The list of switches contains unique values.
2. We check if the link id is a list of monitored links. 

When both conditions are TRUE, we return TRUE.

Reply:
Thanks for clarifying, it looks like there was a bug in the existing implementation, missing the association between the switch and the link.

nit: Changing `IsSwitchWatched(si.EntityId, sysInfo) && si.EntityId == switchID` to `IsSwitchWatched(si.EntityId == switchID && si.EntityId, sysInfo)` can avoid some of the function call.

Reply:
Good point!